lhotse/cut/data.py

```diff
@@ -0,0 +1,960 @@
+import logging
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass, field
+from functools import partial
+from math import isclose
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union
+
+import numpy as np
+import torch
+from intervaltree import IntervalTree
+
+from lhotse.audio import Recording
+from lhotse.augmentation import AugmentFn
+from lhotse.cut.base import Cut
+from lhotse.features import FeatureExtractor, Features
+from lhotse.features.io import FeaturesWriter
+from lhotse.supervision import SupervisionSegment
+from lhotse.utils import (
+    LOG_EPSILON,
+    Pathlike,
+    Seconds,
+    TimeSpan,
+    add_durations,
+    compute_num_frames,
+    compute_num_samples,
+    fastcopy,
+    ifnone,
+    measure_overlap,
+    overlaps,
+    overspans,
+    perturb_num_samples,
+    rich_exception_info,
+    uuid4,
+)
+
+
+@dataclass
+class DataCut(Cut, metaclass=ABCMeta):
```

Comment:
I don't know how you made abstract classes work with dataclasses and inheritance, but you seem to have achieved something I could not :) kudos

Reply:
I think a bit of help from Github copilot XD